src/ray/gcs/tables.cc

```diff
@@ -500,31 +522,61 @@ Status ClientTable::Connect(const ClientTableDataT &local_client) {
                            subscription_callback));
   };
-  return Append(JobID::nil(), client_log_key_, data, add_callback);
+  return Add(JobID::nil(), client_log_key_, data, add_callback);
 }
 
 Status ClientTable::Disconnect(const DisconnectCallback &callback) {
   auto data = std::make_shared<ClientTableDataT>(local_client_);
-  data->is_insertion = false;
-  auto add_callback = [this, callback](AsyncGcsClient *client, const ClientID &id,
-                                       const ClientTableDataT &data) {
+  auto remove_callback = [this, callback](AsyncGcsClient *client, const ClientID &id,
+                                          const ClientTableDataT &data) {
     HandleConnected(client, data);
     RAY_CHECK_OK(CancelNotifications(JobID::nil(), client_log_key_, id));
     if (callback != nullptr) {
       callback();
     }
   };
-  RAY_RETURN_NOT_OK(Append(JobID::nil(), client_log_key_, data, add_callback));
+  auto failure_callback = [](gcs::AsyncGcsClient *client, const UniqueID &id) {
+    RAY_LOG(FATAL) << "This should not happen in ClientTable::Disconnect.";
+  };
+  // Change is_connected of this entry from true to false.
+  auto new_data = std::make_shared<ClientTableDataT>(*data);
+  new_data->is_connected = false;
+  RAY_CHECK_OK(
+      Replace(JobID::nil(), client_log_key_, data, new_data, nullptr, failure_callback));
   // We successfully added the deletion entry. Mark ourselves as disconnected.
   disconnected_ = true;
   return Status::OK();
 }
 
 ray::Status ClientTable::MarkDisconnected(const ClientID &dead_client_id) {
-  auto data = std::make_shared<ClientTableDataT>();
-  data->client_id = dead_client_id.binary();
-  data->is_insertion = false;
-  return Append(JobID::nil(), client_log_key_, data, nullptr);
+  if (client_id_ == dead_client_id) {
+    return Disconnect(nullptr);
+  } else {
+    auto lookup_callback = [this, dead_client_id](
+        AsyncGcsClient *client, const ClientID &id,
+        const std::vector<ClientTableDataT> &clients) {
+      std::shared_ptr<ClientTableDataT> data;
+      for (auto &client_data : clients) {
+        if (client_data.client_id == dead_client_id.binary()) {
+          data = std::make_shared<ClientTableDataT>(client_data);
+          break;
+        }
+      }
+      if (data != nullptr) {
+        // Change is_connected of this entry from true to false.
+        auto failure_callback = [](gcs::AsyncGcsClient *client, const UniqueID &id) {
+          RAY_LOG(FATAL) << "This should not happen in ClientTable::MarkDisconnected.";
+        };
+        auto new_data = std::make_shared<ClientTableDataT>(*data);
+        new_data->is_connected = false;
+        RAY_CHECK_OK(Replace(
+            JobID::nil(), client_log_key_, data, new_data, nullptr, failure_callback));
+      } else {
+        RAY_LOG(WARNING) << "Trying to mark a raylet node that is not registered.";
+      }
+    };
+    return Lookup(lookup_callback);
```

Comment:
What if multiple nodes call `replace` at the same time? Will there be race conditions?

Reply:
Yes, the monitor and the raylet may both try to change the status. I will change the Log level from `FATAL` to `ERROR`.